bpf/nat64.c

```diff
@@ -277,6 +256,38 @@ static __always_inline int nat46(struct __sk_buff *skb)
 	ip6.daddr.in6_u.u6_addr32[2] = 0;
 	ip6.daddr.in6_u.u6_addr32[3] = bpf_htonl(dst_addr);
 
+	// Calculate L4 checksum difference before any helpers that modify
+	// packet's data are called, because verifier will invalidate
+	// all packet pointers.
+	__u64 l4_csum_diff = 0;
+	// Move declarations outside of switch case since declaration
+	// as first instruction in case is illegal in C.
+	const struct udphdr *uh;
+	switch (ip6.nexthdr) {
+		case IPPROTO_UDP:
+			uh = (const struct udphdr *)(ip4 + 1);
+			// If IPv4/UDP checksum is 0 then fallback to clatd so it can calculate the
+			// checksum. Otherwise the network or more likely the NAT64 gateway might
+			// drop the packet because in most cases IPv6/UDP packets with a zero checksum
+			// are invalid. See RFC 6935.
+			if (!uh->check) {
+				bpf_printk("NAT46 packet forwarded: UDP checksum is 0");
+				return TC_ACT_OK;
+			}
+			// If checksum is not 0, recalculate with bpf_csum_diff.
+			// Fallthrough to the next case.
```

Comment:
this is an artifact from my copy paste from the android implementation, 0 checksum in UDP for IPv4 is ok, so we need to recalculate the checksum, This seems to rely on another component called clatd, but we are not using it, so just recalculate if is possible